internal/functions/serve/serve.go

```diff
@@ -59,179 +52,84 @@ func (i *RuntimeOption) toArgs() []string {
-			Env:          env,
-			Entrypoint:   entrypoint,
-			ExposedPorts: exposedPorts,
-			WorkingDir:   utils.ToDockerPath(cwd),
-			// No tcp health check because edge runtime logs them as client connection error
-		},
-		container.HostConfig{
-			Binds:        binds,
-			PortBindings: portBindings,
-		},
-		network.NetworkingConfig{
-			EndpointsConfig: map[string]*network.EndpointSettings{
-				utils.NetId: {
-					Aliases: utils.EdgeRuntimeAliases,
-				},
-			},
-		},
-		utils.EdgeRuntimeId,
-	)
-	return err
-}
 
-func parseEnvFile(envFilePath string, fsys afero.Fs) ([]string, error) {
-	if envFilePath == "" {
-		if f, err := fsys.Stat(utils.FallbackEnvFilePath); err == nil && !f.IsDir() {
-			envFilePath = utils.FallbackEnvFilePath
-		}
-	} else if !filepath.IsAbs(envFilePath) {
-		envFilePath = filepath.Join(utils.CurrentDirAbs, envFilePath)
-	}
-	env := []string{}
-	secrets, err := set.ListSecrets(envFilePath, fsys)
-	for _, v := range secrets {
-		env = append(env, fmt.Sprintf("%s=%s", v.Name, v.Value))
-	}
-	return env, err
-}
+	// Create a channel to signal when a restart is needed
+	restartChan := make(chan struct{})
 
-func populatePerFunctionConfigs(cwd, importMapPath string, noVerifyJWT *bool, fsys afero.Fs) ([]string, string, error) {
-	slugs, err := deploy.GetFunctionSlugs(fsys)
-	if err != nil {
-		return nil, "", err
-	}
-	functionsConfig, err := deploy.GetFunctionConfig(slugs, importMapPath, noVerifyJWT, fsys)
-	if err != nil {
-		return nil, "", err
-	}
-	binds := []string{}
-	for slug, fc := range functionsConfig {
-		if !fc.Enabled {
-			fmt.Fprintln(os.Stderr, "Skipped serving Function:", slug)
-			continue
-		}
-		modules, err := deploy.GetBindMounts(cwd, utils.FunctionsDir, "", fc.Entrypoint, fc.ImportMap, fsys)
-		if err != nil {
-			return nil, "", err
-		}
-		binds = append(binds, modules...)
-		fc.ImportMap = utils.ToDockerPath(fc.ImportMap)
-		fc.Entrypoint = utils.ToDockerPath(fc.Entrypoint)
-		functionsConfig[slug] = fc
-		for i, val := range fc.StaticFiles {
-			fc.StaticFiles[i] = utils.ToDockerPath(val)
+	// Create a map to track watched directories
+	watchedDirs := make(map[string]bool)
+
+	// Start the file watcher in a goroutine
+	go runFileWatcher(ctx, watcher, watchedPath, restartChan, watchedDirs, fsys)
+
+	errChan := make(chan error, 1)
+
+	for {
+		select {
+		case <-ctx.Done():
+			fmt.Println("Stopping functions server...")
+			// 2. Remove existing container if any.
+			_ = utils.Docker.ContainerRemove(context.Background(), utils.EdgeRuntimeId, container.RemoveOptions{
+				RemoveVolumes: true,
+				Force:         true,
+			})
+			return ctx.Err()
+		default:
+			// Use network alias because Deno cannot resolve `_` in hostname
+			dbUrl := fmt.Sprintf("postgresql://postgres:postgres@%s:5432/postgres", utils.DbAliases[0])
+
+			serviceCancel, logsDone, err := manageFunctionServices(ctx, envFilePath, noVerifyJWT, importMapPath, dbUrl, runtimeOption, fsys, errChan)
```

Comment:
Instead of returning a service cancel, can we simply call `Docker.ContainerRestart` on receiving events from the restart channel?

I believe the docker log stream will be disconnected on container restart so we can catch and ignore the error to restream.

Reply:
ContainerRestart won't be enough to update bindings and so on, we need a full container recreation.